### PR TITLE
(#86) Enforce permissions on created files

### LINKED ISSUE #86: Need a way to enforce permissions on created files
For compliance reasons, we need to be able to enforce ownership and permissions on the files and directories created by the module.

### PR DESCRIPTION
**Pull request by silug**
* Add attributes for enforcing ownership and permissions on created files
* Add data types to all class and defined type parameters
* Move documentation to Puppet Strings
* Add generated REFERENCE.md
* Add rspec tests for `kmod::option`, new parameters, etc.
* Bump stdlib dependency to 5.0.0 (for `Stdlib::Filemode` type)

Fixes #86

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -6,19 +6,33 @@
 #
 #
 class kmod (
-  Hash $list_of_aliases    = {},
-  Hash $list_of_blacklists = {},
-  Hash $list_of_installs   = {},
-  Hash $list_of_loads      = {},
-  Hash $list_of_options    = {},
+  Hash             $list_of_aliases,
```
Comment (bastelfreak) [inline manifests/init.pp:9]: please add the default values here and not as data-in-modules. that enabled puppet-strings to pick it up and add it to the REFERENCE.md.

File: manifests/install.pp [hunk 2]
```diff
@@ -16,13 +16,21 @@
   $ensure=present,
   $command='/bin/true',
   $file="/etc/modprobe.d/${name}.conf",
+  Optional[String[1]]        $owner = undef,
```
Comment (bastelfreak) [inline manifests/install.pp:19]: please document all new parameters with the puppet-strings style in the header

File: manifests/init.pp [hunk 3]
```diff
@@ -1,24 +1,52 @@
...
+# @example
+#   include kmod
 class kmod (
-  Hash $list_of_aliases    = {},
-  Hash $list_of_blacklists = {},
-  Hash $list_of_installs   = {},
-  Hash $list_of_loads      = {},
-  Hash $list_of_options    = {},
+  # Defaults for these are in Hiera to enable using them as
+  # defaults in the various defined types via `lookup()`.
```
Comment (bastelfreak) [inline manifests/init.pp:20]: but why do you want to use lookup()? all the defined resources require the kmod class, you could reference the variables from that class.

Comment (silug) [reply manifests/init.pp:20]: I was getting failures from rspec when I tried that.

Comment (silug) [reply manifests/init.pp:20]: Testing again, if I do this:
```diff
diff --git a/manifests/load.pp b/manifests/load.pp
index 6823751..854c9e5 100644
--- a/manifests/load.pp
+++ b/manifests/load.pp
@@ -16,10 +16,10 @@ define kmod::load (
   Stdlib::Unixpath           $file     = '/etc/modules',
   # Defaults for these are set in Hiera so they can be shared with the `kmod` class.
   # lint:ignore:lookup_in_parameter
-  Optional[String[1]]        $owner    = lookup('kmod::owner', String[1], 'first', undef),
-  Optional[String[1]]        $group    = lookup('kmod::group', String[1], 'first', undef),
-  Optional[Stdlib::Filemode] $mode     = lookup('kmod::file_mode', String[1], 'first', undef),
-  Optional[Stdlib::Filemode] $exe_mode = lookup('kmod::exe_mode', String[1], 'first', undef),
+  Optional[String[1]]        $owner    = $kmod::owner,
+  Optional[String[1]]        $group    = $kmod::group,
+  Optional[Stdlib::Filemode] $mode     = $kmod::file_mode,
+  Optional[Stdlib::Filemode] $exe_mode = $kmod::exe_mode,
   # lint:endignore
 ) {
   include kmod
```
I get `error during compilation: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Unknown variable: 'kmod::owner'.`

Alternatively, I could drop the settings on the individual defined types and directly reference `$kmod::owner` (etc.) directly in the defined types, but then we lose the ability to override those settings for the individual types.

Comment (silug) [reply manifests/init.pp]: @bastelfreak I removed the `lookup()` calls (and the ability to override permissions on individual resources).  Please take another look when you have a moment.

File: manifests/install.pp [hunk 4]
```diff
@@ -1,22 +1,18 @@
...
-#
 define kmod::install (
-  $ensure=present,
-  $command='/bin/true',
-  $file="/etc/modprobe.d/${name}.conf",
+  Enum['present', 'absent'] $ensure  = 'present',
+  String[1]                 $command = '/bin/true',
+  Stdlib::Unixpath          $file    = "/etc/modprobe.d/${name}.conf",
 ) {
+  include kmod
```
Comment (bastelfreak) [inline manifests/install.pp:14]: why did you add the include here?

Comment (silug) [reply manifests/install.pp:14]: We need that to make sure that `/etc/modprobe.d` is managed.

**Review by bastelfreak — APPROVED** [review]
